### PR TITLE
feat: add `reload` argument to `*Job.done()` functions

### PR DESCRIPTION
**Pull request by tswast**
This enables checking the job status without making an API call.

It also fixes an inconsistency in `QueryJob`, where a job can be
reported as "done" without having the results of a `getQueryResults` API
call.

Follow-up to https://github.com/googleapis/python-bigquery/pull/340

### CODE REVIEW

File: google/cloud/bigquery/job.py [hunk 1]
```diff
@@ -3082,10 +3085,25 @@ def done(self, retry=DEFAULT_RETRY, timeout=None):
...
+                If ``True``, make an API call to refresh the job state of
+                unfinished jobs before checking. Default ``True``.
 
         Returns:
             bool: True if the job is complete, False otherwise.
         """
+        is_done = (
+            # Only consider a QueryJob complete when we know we have the final
+            # query results available.
+            self._query_results is not None
```
Comment (shollyman) [inline google/cloud/bigquery/job.py:3098]: checking: What about DDL statements, where we have no results?  Or would this already retain the empty row iterator?

**Review by shollyman — APPROVED** [review]
Do the changes here indicate there's further changes for registering the done callback?  My mental model has them independent, but I wanted to make sure.

File: google/cloud/bigquery/job.py (hunk 1 above)
Comment (tswast) [reply google/cloud/bigquery/job.py:3098]: DDL statements don't 404 for `jobs.getQueryResults`, do they? This is the full response object of `jobs.getQueryResults`, not just the rows.

**Comment by tswast** [comment]
> Do the changes here indicate there's further changes for registering the done callback?

No callbacks here. Maybe when we implement the async client https://github.com/googleapis/python-bigquery/issues/18

`done` does get called by `result` via [PollingFuture._blocking_poll](https://github.com/googleapis/python-api-core/blob/85e14e3f3a6b3e3c3ae9626daf60183f4c05fb6a/google/api_core/future/polling.py#L107), which is why I can now remove the extra call to `_get_query_results` from `QueryJob.result()`.

**Comment by tswast** [comment]
> Do the changes here indicate there's further changes for registering the done callback?

Maybe you meant -- How do we get the needed arguments to `get_query_results` from `result`? I'm planning to do something a little bit hacky in https://github.com/googleapis/python-bigquery/pull/347/files -- which maybe is a bad idea, as it means you can't call `result` from two threads simultaneously. Hmmm...

**Pull request merged**